analysis/kb_analysis.py

```diff
@@ -0,0 +1,164 @@
+import datetime
+import json
+import logging
+import math
+import os
+from argparse import ArgumentParser
+from typing import Dict, List, TypedDict
+
+import matplotlib.pyplot as plt
+import numpy as np
+from sklearn.linear_model import LinearRegression
+
+from analysis.config import load_config, Config
+from analysis.shared_parsers import extract_year_ticks, next_year_quarter
+from models.bass_diffusion import BassDiffusionModel
+
+Filetype = str
+PeriodicFiletypeCount = Dict[Filetype, Dict[str, int]]
+
+
+class PeriodCount(TypedDict):
+    period: str
+    count: int
+
+
+SortedFileCount = Dict[Filetype, List[PeriodCount]]
+
+
+def main(config: Config) -> int:
+    start = datetime.datetime.now()
+    kb_cfg = config['data']['kb']
+
+    aggregate_stats_path = os.path.join(kb_cfg['json_output_dir'], 'kb_aggregate_stats.json')
+    with open(aggregate_stats_path, 'rt') as f:
+        aggregate_stats = json.loads(f.read())
+
+    quarterly_counts = to_sorted_quarterly(aggregate_stats)
+    plot_counts(quarterly_counts, kb_cfg)
+
+    end = datetime.datetime.now()
+    logging.info(f'Script took {end - start}')
+
+    return 0
+
+
+def to_sorted_quarterly(file_type_montly_counts: PeriodicFiletypeCount) -> SortedFileCount:
+    quarterly_counts: SortedFileCount = {}
+
+    for file_type, monthly_counts in file_type_montly_counts.items():
+        quarterly_counts.setdefault(file_type, [])
+
+        time_sorted = list(monthly_counts.items())
+        time_sorted = sorted(time_sorted, key=lambda stats: stats[0])
+
+        for year_month, count in time_sorted:
+            year = year_month.split('-')[1]
```

Comment:
Ik liep bij de DANS analyse tegen dezelfde OOM errors aan. Het komt erop neer dat ik de brondata beter moet checken: als het jaar in de toekomst ligt, zal de beruchte while-loop nooit eindigen. Deze twee checks moeten voorkomen dat dit misloopt:

```suggestion
            # Make sure the year/month string is properly formatted
            if re.match(pattern=r'\d{4}-\d{2}', string=year_month) is None:
                logging.warning(f'Expected year-month formatted YYYY-mm, got {year_month}, skipping')
                continue

            # Make sure that the year is not in the future, otherwise the autofill of intermediate
            # Missing 0-count periods until the current period end only until your memory runs out
            year = int(year_month.split('-')[0])
            if year > current_year:
                logging.warning(f'Expected year entry not to be in the future, got {year}, skipping')
                continue

```